staging/src/k8s.io/client-go/openapi3/root.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi3
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/spec3"
+)
+
+const OpenAPIV3Endpoint = "/openapi/v3"
+
+// Root interface defines functions implemented against the root
+// OpenAPI V3 document. The root OpenAPI V3 document maps the
+// API Server relative url for all GroupVersions to the relative
+// url for the OpenAPI relative url. Example for single GroupVersion
+// apps/v1:
+//
+//	"/apis/apps/v1": {
+//	    "ServerRelativeURL": "/openapi/v3/apis/apps/v1?hash=<HASH>"
+//	}
+type Root interface {
+	// GroupVersions returns every GroupVersion for which there is an
+	// OpenAPI V3 GroupVersion document. Returns an error for problems
+	// retrieving or parsing the OpenAPI V3 root document.
+	GroupVersions() ([]schema.GroupVersion, error)
+	// GVSchemas returns the schemas for all the resources in a
+	// GroupVersion as a pointer to a spec3.OpenAPI struct.
+	// Returns an error for problems retrieving or parsing the root
+	// document or GroupVersion OpenAPI V3 document.
+	GVSchemas(gv schema.GroupVersion) (*spec3.OpenAPI, error)
+	// GVSchemasAsBytes returns the schemas for all the resources in a
+	// GroupVersion as unstructured bytes. Returns an error for
+	// problems retrieving or parsing the root or GroupVersion
+	// OpenAPI V3 documents.
+	GVSchemasAsBytes(gv schema.GroupVersion) ([]byte, error)
+}
+
+// openapi3Root implements the Root interface, and encapsulates the
+// fields to retrieve, store the parsed OpenAPI V3 root document.
+type openapi3Root struct {
+	// Sync structure to allow only one download of root document.
+	once sync.Once
+	// REST client to retrieve the OpenAPI V3 documents.
+	client rest.Interface
+	// Parsed root OpenAPI V3 document mapping the GroupVersions
+	// to the OpenAPI V3 relative urls for subsequent retrieval of
+	// the GroupVersion schemas.
+	gvSchemas map[schema.GroupVersion]string
+	// Stores error in retrieval/parsing of root document.
+	err error
+}
+
+// Validate openapi3Root implements the Root interface.
+var _ Root = &openapi3Root{}
+
+// NewRoot returns a structure implementing the Root interface,
+// created with the passed rest client.
+func NewRoot(client rest.Interface) Root {
```

Comment:
This should be built on top of the existing client we already have, that knows how to download the openapi.

Reply:
I initially implemented it that way. Then I realized we must send in a REST client if we expect to use HTTP Etags for disk caching (like aggregated discovery documents).